javascript/src/ExtendedQuery.ts

```diff
@@ -0,0 +1,278 @@
+    assert.ok('Expected PickleStep to have an astNodeId')
+    return this.stepById.get(astNodeId)
+  }
+
+  findPickleBy(testCaseStarted: TestCaseStarted) {
+    const testCase = this.findTestCaseBy(testCaseStarted)
+    assert.ok(testCase, 'Expected to find TestCase from TestCaseStarted')
+    return this.pickleById.get(testCase.pickleId)
+  }
+
+  findPickleStepBy(testStep: TestStep) {
+    assert.ok(testStep.pickleStepId, 'Expected TestStep to have a pickleStepId')
+    return this.pickleStepById.get(testStep.pickleStepId)
+  }
+
+  findAllTestCaseStarted(): ReadonlyArray<TestCaseStarted> {
+    return [...this.testCaseStarted]
+  }
+
+  findTestCaseBy(testCaseStarted: TestCaseStarted) {
+    return this.testCaseById.get(testCaseStarted.testCaseId)
+  }
+
+  findTestCaseFinishedBy(testCaseStarted: TestCaseStarted) {
+    return this.testCaseFinishedByTestCaseStartedId.get(testCaseStarted.id)
+  }
+
+  findTestCaseDurationBy(testCaseStarted: TestCaseStarted) {
+    const testCaseFinished = this.findTestCaseFinishedBy(testCaseStarted)
+    if (!testCaseFinished) {
+      return undefined
+    }
+    return TimeConversion.millisecondsToDuration(
+      TimeConversion.timestampToMillisecondsSinceEpoch(testCaseFinished.timestamp) -
+        TimeConversion.timestampToMillisecondsSinceEpoch(testCaseStarted.timestamp)
+    )
+  }
+
+  findMostSevereTestStepResultBy(testCaseStarted: TestCaseStarted) {
+    return getWorstTestStepResult(
+      this.findTestStepFinishedAndTestStepBy(testCaseStarted).map(
+        ([testStepFinished]) => testStepFinished.testStepResult
+      )
+    )
+  }
+
+  findTestStepBy(testStepFinished: TestStepFinished) {
+    return this.testStepById.get(testStepFinished.testStepId)
+  }
+
+  findTestStepFinishedAndTestStepBy(
+    testCaseStarted: TestCaseStarted
+  ): ReadonlyArray<[TestStepFinished, TestStep]> {
+    return this.testStepFinishedByTestCaseStartedId
+      .get(testCaseStarted.id)
+      .map((testStepFinished) => {
+        const testStep = this.findTestStepBy(testStepFinished)
+        assert.ok(testStep, 'Expected to find TestStep by TestStepFinished')
+        return [testStepFinished, testStep]
+      })
+  }
+
+  findTestRunDuration() {
+    if (!this.testRunStarted || !this.testRunFinished) {
+      return undefined
+    }
+    return TimeConversion.millisecondsToDuration(
+      TimeConversion.timestampToMillisecondsSinceEpoch(this.testRunFinished.timestamp) -
+        TimeConversion.timestampToMillisecondsSinceEpoch(this.testRunStarted.timestamp)
+    )
+  }
+
+  countMostSevereTestStepResultStatus(): Record<TestStepResultStatus, number> {
+    const result: Record<TestStepResultStatus, number> = {
+      [TestStepResultStatus.AMBIGUOUS]: 0,
+      [TestStepResultStatus.FAILED]: 0,
+      [TestStepResultStatus.PASSED]: 0,
+      [TestStepResultStatus.PENDING]: 0,
+      [TestStepResultStatus.SKIPPED]: 0,
+      [TestStepResultStatus.UNDEFINED]: 0,
+      [TestStepResultStatus.UNKNOWN]: 0,
+    }
```

Comment:
Note: This differs from the Java implementation which only includes results > 0. 

I'm okay with changing that in Query.  Instead of returning a map we can return an `EnumMap` in `query` and remove the `getOrDefault` from Java.